prow/config/config.go

```diff
@@ -1243,25 +1262,116 @@ type Branding struct {
+	// ProwJob is configured to run on. Recall that ProwJobs default to running on
+	// the "default" build cluster if they omit the "cluster" field in config.
+	Cluster string `json:"cluster,omitempty"`
+
+	// Config is the RerunAuthConfig to apply if the filter fields all match the
+	// ProwJob. Note that when multiple entries match a ProwJob the entry with the
+	// highest specification is used.
+	Config *prowapi.RerunAuthConfig `json:"rerun_auth_configs,omitempty"`
+}
+
+func (d *Deck) GetRerunAuthConfig(refs *prowapi.Refs, cluster string) *prowapi.RerunAuthConfig {
+	var configs []*prowapi.RerunAuthConfig
+
+	if refs != nil {
+		for _, drac := range d.DefaultRerunAuthConfigs {
+			if drac.matches(refs.OrgRepoString(), cluster) {
+				configs = append(configs, drac.Config)
+			}
+		}
+	} else {
+		// Wilcard case: add last config
+		dracLength := len(d.DefaultRerunAuthConfigs)
+		if dracLength != 0 {
+			drac := d.DefaultRerunAuthConfigs[dracLength-1]
+			configs = append(configs, drac.Config)
+		}
 	}
 
-	if rerun, exists := rac[fmt.Sprintf("%s/%s", refs.Org, refs.Repo)]; exists {
-		return rerun
+	if len(configs) == 0 {
+		configs = append(configs, &prowapi.RerunAuthConfig{})
 	}
 
-	if rerun, exists := rac[refs.Org]; exists {
-		return rerun
+	return configs[len(configs)-1]
+}
+
+// defaultRerunAuthMapToSlice converts the old format
+// (map[string]*prowapi.RerunAuthConfig) to the new format
+// ([]*DefaultRerunAuthConfigEntry) or DefaultRerunAuthConfigs.
+func defaultRerunAuthMapToSlice(m map[string]prowapi.RerunAuthConfig) DefaultRerunAuthConfigs {
+	var entries DefaultRerunAuthConfigs
+	add := func(repo string, rac prowapi.RerunAuthConfig) {
+		entries = append(entries, &DefaultRerunAuthConfigEntry{
+			OrgRepo: repo,
+			Cluster: "",
+			Config:  &rac,
+		})
 	}
 
-	return rac["*"]
+	// Ensure "*" comes first...
+	if rac, exists := m["*"]; exists {
+		add("*", rac)
+	}
+	// then orgs...
+	for key, rac := range m {
+		if key == "*" || strings.Contains(key, "/") {
+			continue
+		}
+		add(key, rac)
+	}
+	// then repos.
+	for key, rac := range m {
+		if key == "*" || !strings.Contains(key, "/") {
+			continue
+		}
+		add(key, rac)
+	}
+
+	return entries
+}
+
+// FinalizeDefaultRerunAuthConfigs prepares the entries of
+// Deck.DefaultRerunAuthConfigs for use finalizing job config.
+// It parses either d.RerunAuthConfigs or d.DefaultRerunAuthConfigEntries, not both.
+// Old format: map[string]*prowapi.RerunAuthConfig where the key is org,
+//             org/repo, or "*".
+// New format: []*DefaultRerunAuthConfigEntry
+// If the old format is parsed it is converted to the new format, then all
+// filter regexp are compiled.
+func (d *Deck) FinalizeDefaultRerunAuthConfigs() error {
+	mapped, sliced := len(d.RerunAuthConfigs) > 0, len(d.DefaultRerunAuthConfigs) > 0
+
+	if mapped && sliced {
+		return fmt.Errorf("deck.rerun_auth_configs and deck.default_rerun_auth_config_entries are mutually exclusive, please use one or the other")
```

Comment:
Suggest a comment here saying that this should be guarded against by prow config test, the checking here is for cases where prow config test didn't run

Reply:
Comment was added